With Product Variations HOC > Convert to TypeScript

thealexandrelara opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->
This PR converts the Product Variations HOC from JavaScript to TypeScript.

Closes #42361 .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Login to the dashboard.
2. Create a new post (Posts > Add new post)
3. In the editor, you should see a block inserter icon or an option to "Add Block". Click on it.
4. In the block inserter, search for "Reviews" or browse through the available blocks until you find it.
    ![CleanShot 2024-02-14 at 17 37 01@2x](https://github.com/woocommerce/woocommerce/assets/20469356/f40ddfdf-cc93-4a6f-ba51-8a1a65c6c76c)
6. Click on the "Reviews by Product" blocks to add it to your content.
7. Using the Product selector, select a product **without** reviews.
8. Make sure everything is working correctly.
9. Select a product **with** reviews using the Product selector.
10. Make sure everything is working correctly.
11. Save the post and visit it on the frontend to ensure they are working as expected.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->
Convert With Product Variations HOC to TypeScript

</details>

thealexandrelara committed:
Add with-product-variations.tsx

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

thealexandrelara committed:
Add types for ProductVariationItem

thealexandrelara committed:
Fix ESLint errors

thealexandrelara committed:
Remove JS file

samueljseay commented on plugins/woocommerce-blocks/assets/js/hocs/with-product-variations.tsx:
```diff
@@ -3,29 +3,56 @@
  */
 import { formatError } from '../base/utils/errors';
 
+interface Product {
```
A type already exists that is used for other wrapped components called [ProductResponseItem](https://github.com/woocommerce/woocommerce/blob/9203f9233da5bafb312e797e9fa2da063a3086c1/plugins/woocommerce-blocks/assets/js/types/type-defs/product-response.ts#L59) I think its the same type as what you want here isn't it?

thealexandrelara replied:
Nice catch! I forgot to check if there was an already defined type for it. (included in https://github.com/woocommerce/woocommerce/pull/44785/commits/2da39e281430405bf55cb191cb55b0c2cea67fd8)

samueljseay commented on plugins/woocommerce-blocks/assets/js/hocs/with-product-variations.tsx:
```diff
@@ -46,7 +73,7 @@ const withProductVariations = createHigherOrderComponent(
 			}
 
-			loadVariations = () => {
+			loadVariations = (): void => {
```
Any reason why you felt you needed to add this? Its already inferred by TS and this isn't a public API so in my opinion no need to add this.

thealexandrelara replied:
Actually, Copilot added this automatically 😅, I think it is not necessary in this case, so I removed in https://github.com/woocommerce/woocommerce/pull/44785/commits/437ddea1dec583d3b47c453e9ad47fbc7911a30f

samueljseay commented on plugins/woocommerce-blocks/assets/js/hocs/with-product-variations.tsx:
```diff
@@ -81,7 +108,11 @@ const withProductVariations = createHigherOrderComponent(
 
-				getProductVariations( expandedProduct )
+				(
+					getProductVariations( expandedProduct ) as Promise<
```
Not saying you have to do this, but one way you could avoid the cast here is to start moving these utils across to ts separately too. For example `getProductVariations` is in a `utils/index.js` file, you could separate it out to a TS file, pass `ProductResponseVariationsItem[]` to `apiFetch` generic arg and then that util would be able to be correctly typed by everyone without casting which would be preferable.

thealexandrelara replied:
This is a great suggestion! I tried to keep the changes inside this file because we have many issues created for each file that we need to convert to TS. I believe we can open a new issue to work on this `utils/index.js` file since the functions exported from it can also impact other files. WDYT?

samueljseay replied:
Yep I'm fine with that because this wouldn't be a big job, so happy to keep it to another PR

samueljseay commented on plugins/woocommerce-blocks/assets/js/hocs/with-product-variations.tsx:
```diff
@@ -112,12 +143,12 @@ const withProductVariations = createHigherOrderComponent(
 			};
 
-			isProductId( itemId ) {
+			isProductId( itemId: number ): boolean {
```
again, no need to add return type. TS infers it correctly.

samueljseay reviewed (commented):
Looking good to me, I have some suggestions. let me know what you think 😄

thealexandrelara committed:
Replace Product with ProductResponseItem interface

Initially, a Product interface was created to add the required types; however,
it was pointed out that the @woocommerce/types alias already provided an interface
for products (the ProductResponseItem interface).

thealexandrelara committed:
Remove returning types definition

samueljseay reviewed (approved):
This is looking good to me with the new changes 🚢
